src/backend/cdb/partitionselection.c

```diff
@@ -109,25 +108,31 @@ partition_selection(PartitionNode *pn, PartitionAccessMethods *accessMethods, Oi
 
 	Assert(0 < partAttno);
 
-	Relation	rel = relation_open(root_oid, NoLock);
-	TupleDesc	tupDesc = RelationGetDescr(rel);
-
-	Assert(tupDesc->natts >= partAttno);
-
 	int			i;
 	Datum	   *values = palloc0(partAttno * sizeof(Datum));
 	bool	   *isnull = palloc(partAttno * sizeof(bool));
+	TupleDesc	tupDesc = CreateTemplateTupleDesc(partAttno, false);
 
 	for (i = 0; i < partAttno - 1; i++)
+	{
 		isnull[i] = true;
+
+		/*
+		 * Build tuple descriptor with dummy prefix columns that will not be
+		 * used further in the search of partitions
+		 */
+		TupleDescInitEntry(tupDesc, i+1, NULL, BOOLOID, -1, 0);
+	}
 	isnull[partAttno - 1] = isNull;
 	values[partAttno - 1] = value;
+	TupleDescInitEntry(tupDesc, partAttno, NULL, exprTypid, -1, 0);
 
-	PartitionRule *result = get_next_level_matched_partition(pn, values, isnull, tupDesc, accessMethods, exprTypid);
+	PartitionRule *result = get_next_level_matched_partition(pn, values, isnull,
+			tupDesc, accessMethods);
 
+	FreeTupleDesc(tupDesc);
```

Comment:
As I got from https://github.com/greenplum-db/gpdb/pull/11826 we fix the memory leak here.